cookbooks/bcpc/recipes/networking.rb

```diff
@@ -123,105 +131,93 @@
-    %w{ management storage floating }.each do |iface|
-      bash "#{iface} up" do
-        user "root"
-        code <<-EOH
-          ifup #{node[:bcpc][iface][:interface]}
-        EOH
-        not_if "ip link show up | grep #{node[:bcpc][iface][:interface]}"
-      end
-    end
-    
-    bash "routing-management" do
-        user "root"
-        code "echo '1 mgmt' >> /etc/iproute2/rt_tables"
-        not_if "grep -e '^1 mgmt' /etc/iproute2/rt_tables"
-    end
-    
-    bash "routing-storage" do
-        user "root"
-        code "echo '2 storage' >> /etc/iproute2/rt_tables"
-        not_if "grep -e '^2 storage' /etc/iproute2/rt_tables"
-    end
-    
-    template "/etc/network/if-up.d/bcpc-routing" do
-        mode 00775
-        source "bcpc-routing.erb"
-        notifies :run, "execute[run-routing-script-once]", :immediately
-    end
-    
-    execute "run-routing-script-once" do
-        action :nothing
-        command "/etc/network/if-up.d/bcpc-routing"
-    end
-    
-    bash "disable-noninteractive-pam-logging" do
-        user "root"
-        code "sed --in-place 's/^\\(session\\s*required\\s*pam_unix.so\\)/#\\1/' /etc/pam.d/common-session-noninteractive"
-        only_if "grep -e '^session\\s*required\\s*pam_unix.so' /etc/pam.d/common-session-noninteractive"
-    end
+# set up the DNS resolvers
+# we want the VIP which will be running powerdns to be first on the list
+# but the first entry in our master list is also the only one in pdns,
+# so make that the last entry to minimize double failures when upstream dies.
+resolvers=node[:bcpc][:dns_servers].dup
+if node[:bcpc][:management][:vip] and get_nodes_for("powerdns").length() > 0
+  resolvers.push resolvers.shift
+  resolvers.unshift node[:bcpc][:management][:vip]
+end
+
+bash "update resolvers" do
+  code <<-EOH
+  echo "#{(resolvers.map{|r| "nameserver #{r}"} + ["search #{node[:bcpc][:domain_name]}"]).join('\n')}" | resolvconf -a #{node[:bcpc][:management][:interface]}.inet
+  EOH
+end
+
+ifaces = %w(management storage floating)
+ifaces.each_index do |i|
+  iface = ifaces[i]
+  device_name = node[:bcpc][iface][:interface]
+  next if iface != "management" and \
+      node[:bcpc][:management][:interface] == device_name
+  template "/etc/network/interfaces.d/#{device_name}.cfg" do
+    source "network.iface.erb"
+    owner "root"
+    group "root"
+    mode 00644
+    variables(
+      :interface => node[:bcpc][iface][:interface],
+      :ip => node[:bcpc][iface][:ip],
+      :netmask => node[:bcpc][iface][:netmask],
+      :gateway => node[:bcpc][iface][:gateway],
+      :slaves => node[:bcpc][iface].has_key("slaves") ? node[:bcpc][iface][:slaves] : False,
+      :dns => resolvers,
+      :mtu => node[:bcpc][iface][:mtu],
+      :metric => i*100
+    )
+  end
+  
+  bash "#{iface} up" do
+    code <<-EOH
+      ifup #{device_name} #{node[:bcpc][iface].has_key("slaves") and node[:bcpc][iface][:slaves].join(" ")}
+    EOH
+    not_if "ip link show up | grep #{device_name}"
```

Comment:
One thing to be aware of, if one has a change to an `/etc/network/interfaces.d/*` file, this `ifup` will only run if the device was down. If the device was reconfigured (e.g. new MTU) then that change won't take affect until something external (e.g. `ifdown`/`ifup` or a reboot).